controller/pb/progress.go

```diff
@@ -0,0 +1,122 @@
+package pb
+
+import (
+	"github.com/docker/buildx/util/progress"
+	control "github.com/moby/buildkit/api/services/control"
+	"github.com/moby/buildkit/client"
+	"github.com/opencontainers/go-digest"
+)
+
+type writer struct {
+	ch chan<- *StatusResponse
+}
+
+func NewProgressWriter(ch chan<- *StatusResponse) progress.Writer {
+	return &writer{ch: ch}
+}
+
+func (w *writer) Write(status *client.SolveStatus) {
+	w.ch <- ToControlStatus(status)
+}
+
+func (w *writer) ValidateLogSource(digest.Digest, interface{}) bool {
+	return true
+}
+
+func (w *writer) ClearLogSource(interface{}) {}
```

Comment:
I think we should be fine actually

https://github.com/docker/buildx/blob/48b733d6da24265d75f134b6934d67bc657073b5/util/progress/writer.go#L48-L67

They are used by the printer, and since the networked progress writer forwards all data into a printer, we'll actually reuse the logic from there (we might end up doing unnecessary network requests, but it's localhost, so I think it should be fine).